### PR TITLE
FIX: restore creating new axes via plt.subplot with different kwargs

### LINKED ISSUE #10700: gca(projection="rectilinear") fails to reset projection on axes
### Bug report

**Bug summary**

As of 2.2:
```
plt.gca(projection="polar"); plt.gca(projection="lambert")
```
shows a warning and replaces the originally created polar axes by a lambert axes (not sure what behavior we actually want this to have once axes-collision detection in add_axes is removed).
However, in
```
plt.gca(projection="polar"); plt.gca(projection="rectilinear")
```
the second call to gca (using a rectilinear projection) fails to reset the projection at all and leaves the original polar axes in place.

Noted while thinking about #10674.

### LINKED ISSUE #19432: Unexpected change in behavior in plt.subplot
### Bug report

**Bug summary**

#19153 introduced some changes that had (I think) unintended consequences when passing custom kwargs to projections

**Code for reproduction**

Prior to #19153, the following code returned ``False``, that is, two different axes were created:

```python
from astropy.wcs import WCS
import matplotlib.pyplot as plt

wcs1 = WCS(naxis=3)
wcs1.wcs.ctype = ['x', 'y', 'z']

ax1 = plt.subplot(1, 1, 1, projection=wcs1, slices=('x', 'y', 1))

wcs2 = WCS(naxis=3)
wcs2.wcs.ctype = ['a', 'b', 'c']

ax2 = plt.subplot(1, 1, 1, projection=wcs2, slices=('x', 2, 'y'))

print(ax1 is ax2)
```

No deprecation warning or future warning was raised in this case.

Following #19153, this silently ignores the projection and slices in the second call and the print statement returns ``True``.

I don't think #19153 was intended to start making ``plt.subplot`` return the same axes in some cases, rather I thought the goal of that PR was to remove re-use of axes in certain circumstances, so I think this is a regression/bug

**Actual outcome**

``ax1`` and ``ax2`` are the same

**Expected outcome**

``ax1`` and ``ax2`` should be different

### PR DESCRIPTION
**Pull request by tacaswell**
This adds a small amount of additional state to the Axes created via
Figure.add_axes and Figure.add_subplot (which the other Axes creation
methods eventually funnel through) to track the kwargs passed
through.

We then use that state in `pyplot.subplot` to determine if we should
re-use an Axes found at a given position or create a new one (and
implicitly destroy the existing one).

closes #19432
closes #10700

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Review by jklymak — DISMISSED** [review]

**Comment by tacaswell** [comment]
If people are happy with this (which seems to be the case) I still need to chase through the docs to make sure they are still correct.

File: lib/matplotlib/tests/test_figure.py [hunk 1]
```diff
@@ -1010,18 +1010,27 @@ def test_axes_kwargs():
     plt.close()
 
     # plt.subplot() searches for axes with the same subplot spec, and if one
-    # exists, returns it, regardless of whether the axes kwargs were the same.
+    # exists, and the kwargs match returns it, create a new one if they do not
     fig = plt.figure()
     ax = plt.subplot(1, 2, 1)
     ax1 = plt.subplot(1, 2, 1)
     ax2 = plt.subplot(1, 2, 2)
+    # This will delete ax / ax1 as they fully overlap
```
Comment (dstansby) [inline lib/matplotlib/tests/test_figure.py:1018]: Isn't the reason for deletion actually that the projection keywords are not the same?

**Review by dstansby — COMMENTED** [review]
Looks good to me 👍

**Comment by anntzer** [comment]
I believe this suffers from the same underlying problem that caused the old implementation to be so roundabout: some keys are not hashable because they are mutable.  To be concrete, if you write e.g.
```
lims = [.1, .1, .9, .9]
ax1 = fig.add_axes(lims)
lims[0] = .2
ax2 = fig.add_axes(lims)
ax3 = fig.add_axes([.1, .1, .9, .9])
ax4 = fig.add_axes([.2, .1, .9, .9])
```
what are the axes that you think should be the same?  (A similar argument applies e.g. for `add_subplot(111, projection=some_mutable_thing)`.

File: lib/matplotlib/tests/test_figure.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/tests/test_figure.py:1018]: That is what I thought too, but the logic is https://github.com/matplotlib/matplotlib/blob/ea6803254567eaaa9adb027b541a6885d9fa246f/lib/matplotlib/pyplot.py#L1236-L1244 which says "delete any axes which the now axes fully covers" .

**Comment by jklymak** [comment]
@anntzer, I'm not following - first this PR doesn't do anything for add_axes etc. it only affects `plt.subplot`, and the only danger of 
```
plt.subplot(projection=myMutableProjection)
myMutableProjection.mutate()
plt.subplot(projection=myMutableProjection)
``` 
is that they will lose the original plot because the projection has changed, in which case they should just keep the reference:
```
ax = plt.subplot(projection=myMutableProjection)
myMutableProjection.mutate()
plt.sca(ax)
``` 

That seems pretty esoteric to me.  But it is certainly better than breaking the people who expect the axes to clear when doing

```
plt.subplot(projection=proj1)
...
plt.subplot(projection=proj2)
```

**Comment by anntzer** [comment]
I wrote the example with add_axes for simplicity, but again something similar can happen with projections if they are not immutable.  I don't actually know much about cartopy or WCSAxes, but I don't think it would be incredible to have some projections library that supports e.g. `proj.set_central_longitude(whatever)` (for example), in which case you run into the same issue (`proj = Proj(); subplot(proj); <sometime later>; proj.set_central_longitude(whatever); subplot(proj)` which doesn't seem so unlikely to happen).

And if the project is actually immutable, then it should normally be hashable too, so that would be one thing to check.

**Comment by jklymak** [comment]
`proj = Proj(); subplot(proj); <sometime later>; proj.set_central_longitude(whatever); subplot(proj)`

right, and with this change, the second subplot will overwrite the first.  But a) thats hopefully rare, and b) when that happens, presumably its pretty obvious, and we can just tell people to save the reference to the subplot, rather than fragile grabbing from the `subplot` stack

**Comment by anntzer** [comment]
> the second subplot will overwrite the first

I believe that what will happen is that the second call will return the preexisting axes as is?

> presumably its pretty obvious, and we can just tell people to save the reference to the subplot

Just to be clear, it is far from obvious to me what people using the implicit machinery may be expecting, because it's been too long since I've last used it for this kind of manipulations, but heh.

**Comment by tacaswell** [comment]
The key behavior change in #19153 (and what caused the re-write from #18978 is that we moved the re-use or new (and remove) logic into pyplot from the `Figure` methods.  I think everyone is still on board with `add_subplot` doing what it says on the tin and always adding a new Axes to the Figure.   I think it is fair to ask people who are using the explicit (nee OO) version of the API to explicitly keep track of their Axes(es) and the only behavior still up for discussion is what `pyplot` does.  

The argument for keeping the "create or query" logic in pyplot is that if the users are going fully down the implicit route (a number of years ago I had an lengthy discussion with a user who was convinced beyond doubt that the implicit API was easier and clearer to use, 30-45 minutes of effort and several drinks had no effect) then this is the only way for them to come back to an Axes via the internal `sca` call (which I discovered while writing this up was broken so fixed that too).

------

If we have a well behaved non-mutable input (or the users just never mutate it ;) ) we have:

```
obj = SomeClass()
ax1 = plt.subplot(111, keyword=obj)
obj2 = SomeClass()
ax2 = plt.subplots(111, keyword=obj2)
```

- on 3.3 this returns a new Axes and removes the existing one
- on master this return the first Axes because we ignore the kwargs when doing the matching  (which @astrofrog makes a good case is a major API change, I think we have to back out of one way or the other)
- with this patch it return a new Axes and remove the existing one.

However @anntzer is correct that this is opening is up to some new odd behavior when the values are mutable.  Python is too flexible to reliably tell if something is mutable (and even checking if it is hashable is not technically enough as you can over-ride `__hash__` is questionable ways).

```
obj = MutableClass()
ax1 = plt.subplot(111, keyword=obj)
obj.mutate()
ax2 = plt.subplots(111, keyword=obj)
```

 - On 3.3, this gives you back the same Axes or fail to create the first axes depending on if `obj` is hashable.
 - On master we fully ignore the kwarg the second time through and return the existing axes
 - With this patch we still get back the same Axes because we decide that the kwargs are equal

Reusing the axes may or may not be the right thing because it will end up depending on how the projections work (which is between the objects being based in and the project code we have discovered).  If they look at what ever state is in `obj` and draw time, then getting back the same Axes is correct, if the state is used at Axes creation time then this behavior is wrong.

We could make a shallow copy of all of the values, but that would only push the problem "down a layer" or make things wrong in the opposite case as above.  We could make a deep copy which may be expensive and inverts the cases where the behavior is wrong.

There is already going to be an issue the values are numpy arrays ("truth value of an array is ambiguous!"), but they also didn't work in <=3.3 so I am not worried about that 

I think on net, the behavior in this patch is the best option (modulo adding some docs).  In addition we could:
 - always warn in pyplot with kwargs, but that would be annoying to polar or 3D users (who are giving us string values).
 - explicitly allow some known OK keys / values and warn on anything else
 - try to hash the values as a hokey test

I'm inclined to leave it as-is and not get too deep on the semantics of this behavior with the "fix" for user problems is "please use the explicit API where you have a variable name to keep track of what Axes you want to use rather than a proxy variable name through some shared state".

**Comment by jklymak** [comment]
OK, I misunderstood @anntzer.  The problem is that the user might mutate the object and think that they should get a new axes when they call subplot the second time.  I guess I am OK with that being squishy and I still stand by my approval of this approach...

**Comment by tacaswell** [comment]
I am now a _bit_ worried about `pyplot.axes` as well.

**Comment by anntzer** [comment]
Per the above I don't really know what's best for plt.subplot() users, so as long as the implications with mutable args have been thought out I'm fine with that (I agree with @tacaswell fig.add_subplot should just always add a new axes).

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -576,12 +576,12 @@ def add_axes(self, *args, **kwargs):
             if not np.isfinite(rect).all():
                 raise ValueError('all entries in rect must be finite '
                                  'not {}'.format(rect))
-            projection_class, kwargs = self._process_projection_requirements(
```
Comment (lpsinger) [inline lib/matplotlib/figure.py:579]: Why the change in identifier here?

Comment (lpsinger) [reply lib/matplotlib/figure.py:579]: I think that this should be changed back to `kwargs`.

**Comment by lpsinger** [comment]
I have confirmed that, if merged, this will fix the issues in Astropy. See astropy/astropy#11315.

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (jklymak) [reply lib/matplotlib/figure.py:579]: Well we are going to strip some kwargs, so it makes sense to return something different here I think?

Comment (lpsinger) [reply lib/matplotlib/figure.py:579]: The stripping has already been done above, in this function. This call to `_process_projection_requirements` _normalizes_ the remaining keyword arguments. You probably want to save the _normalized_ keyword arguments. Changing this identifier actually changes the behavior of the code in a very minor but probably undesirable way.

**Comment by astrofrog** [comment]
This looks good to me, thanks @tacaswell!

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/figure.py:579]: The issue is that we need to do the check in pyplot on the stashed values to determine if we need to enter this function or not so I think we need to cache the un-processed kwargs (which is what we were doing for 3.3 and prior).

There is the issue that we pop `'rect'` out of the kwargs, however `plt.subplot` forces `args` to be non-empty and `add_axes` will raise if you pass both 'rect' and non-empty args so we can never drive to this case.

Comment (lpsinger) [reply lib/matplotlib/figure.py:579]: By "this function," do you mean `_process_projection_requirements` or `_add_axes_internal`?

Comment (tacaswell) [reply lib/matplotlib/figure.py:579]: I mean `Figure.add_axes`.  The only place we have this state is in `plt.subplot` where the logic is:

 - use args to compute _positional_ duplication
 - if one is found, check the kwargs
   - if user passed kwargs match the stashed kwargs, use them
   - if they do not match, make a new one via `Figure.add_axes` and stash the new user passed kwargs.

Comment (lpsinger) [reply lib/matplotlib/figure.py:579]: Oh, I understand now.

Note, the logic that maps `polar=True` and `projection='polar'` to the same key lives in `_process_projection_requirements`. So in the past, these two statements would have returned the same axes:

```python
ax1 = plt.subplot(111, projection='polar')
ax2 = plt.subplot(111, polar=True)
assert ax1 is ax2  # passes for Matplotlib 3.3.4, but will not pass with this patch
```

I think that this normalization logic has to be lifted out into `plt.subplot` if we want this to work still in this one special case.

File: lib/matplotlib/tests/test_figure.py (hunk 1 above)
Comment (lpsinger) [reply lib/matplotlib/tests/test_figure.py:1018]: @tacaswell is correct. I wasn't aware of this behavior, but it's well documented, and it's been around a long time, so this is 👍 for me.

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/figure.py:579]: drat :(

Comment (tacaswell) [reply lib/matplotlib/figure.py:579]: Pushed new changes so that we now track the projection class + post processed kwargs.  Added @lpsinger 's code as a tost case (thank you :) ).

Comment (lpsinger) [reply lib/matplotlib/figure.py:579]: Looks good now!

Comment (jklymak) [reply lib/matplotlib/figure.py:579]: The point of keeping `plt.subplot` ability to grab the old axes was to give pyplot users a way to get the handle that they were otherwise lacking.   I am not convinced that adding extra logic to handle the above case was a good idea in order to support retrieving the axes.

Comment (lpsinger) [reply lib/matplotlib/figure.py:579]: The new version calls `_process_projection_requirements` twice: once in `plt.subplot`, and then again in `Figure.add_subplot`, which is called by the former. As long as we are OK with that very minor inefficiency, this looks good.

**Review by lpsinger — APPROVED** [review]

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (jklymak) [reply lib/matplotlib/figure.py:579]: My concern is that the previous version was simpler (at least to me), and I don't know know that we need to support recovering the axes using two different spellings for polar.  Overall our code base has lots of clever things in it that support a bunch of weird edge cases like this that makes it very difficult to follow what the code is doing.  If we can avoid that as much as is reasonable, it would be great.

**Comment by tacaswell** [comment]
Sigh, the polar stuff is worse than I thought, the curent code breaks that in a _new_ way, hope to have this sorted in the next few hours.

**Comment by anntzer** [comment]
See also #10700.

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/figure.py:579]: I don't see anyway around the inefficiency (short of some pretty major refactors).

I agree it is annoying we have 2 ways of spelling polar, but it used to work and "it is annoying to normalize" is not a compelling reason to break back-compatibility.

The final version of this is going to be a fair bit more complicated, but at least all of that complication is in `pyplot` rather than spread through the `Figure` codebase.

**Comment by tacaswell** [comment]
OK, _this_ time I think I got enough of the corner cases.  It is a fair bit of code, but I think we have gotten all of the logic layed out in one place now :)  Most of the line count is additional tests.

**Review by QuLogic — APPROVED** [review]

File: doc/api/next_api_changes/behavior/19438-TAC.rst [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+The first time `.pyplot.subplot` is called for a given position in the
+Axes grid it is clear that the correct behavior is to create and
+return an Axes at the given position with the passed arguments
+(defaulting to a rectilinear Axes).  However, on subsequent calls to
+`.pyplot.subplot` the correct behavior is more subtle.  If an Axes exists
+at the requested location and has equivalent parameters the existing Axes
+should be returned, but if the parameters are different the old Axes is removed
+from the Figure and a new Axes is created and returned.
+
+Due to an implementation detail, it was previously the case that for all
```
Comment (lpsinger) [inline doc/api/next_api_changes/behavior/19438-TAC.rst:19]: Is this implementation detail what was introduced by my changes in #19153? If so, we need not say that it was "previously the case" because those changes have never been in a release yet.

File: lib/matplotlib/pyplot.py [hunk 4]
```diff
@@ -1199,7 +1200,27 @@ def subplot(*args, **kwargs):
...
+
     """
+    proj_placeholder = object()
+    # sort out if the user passed us either polar or projection
+    # in principle we need to handle if *axes_class* is passed, but
+    # we will leave that validation to down stream code.  Here we will
+    # only normalize `polar=True` vs `projection='polar'` so we can decide
+    # if we should create a new Axes or return the existing one
+    projection = kwargs.get('projection', proj_placeholder)
+    polar = kwargs.pop('polar', proj_placeholder)
```
Comment (lpsinger) [inline lib/matplotlib/pyplot.py:1215]: Is it really OK to modify `kwargs` here? If the user has called `plt.subplot(111, **kwargs)`, then they may be surprised that kwargs is modified.

File: lib/matplotlib/pyplot.py [hunk 5]
```diff
@@ -1199,7 +1200,27 @@ def subplot(*args, **kwargs):
 
         # add ax2 to the figure again
         plt.subplot(ax2)
+
+        # make the first axes "current" again
+        plt.subplot(221)
+
     """
+    proj_placeholder = object()
```
Comment (lpsinger) [inline lib/matplotlib/pyplot.py:1208]: Why is the placeholder object necessary? Why not `None`?

Comment (anntzer) [inline lib/matplotlib/pyplot.py:1208]: perhaps call this `unset`, I guess the conditions below read better?

**Review by lpsinger — APPROVED** [review]

File: lib/matplotlib/pyplot.py [hunk 6]
```diff
@@ -1228,11 +1249,29 @@ def subplot(*args, **kwargs):
         (ax for ax in fig.axes
          if hasattr(ax, 'get_subplotspec') and ax.get_subplotspec() == key),
         None)
+    # if we found an axes at the right position sort out if we
```
Comment (anntzer) [inline lib/matplotlib/pyplot.py:1252]: the `ax = ax` below are a bit weird even though I get the logic.  Perhaps rewrite starting by removing the call to `next()` and then
```
for ax in fix.axes:
    if hasattr(ax, "...") and ax.get_subplotspec() == key:
        p_class, pkw = ...
        if <reuse>:
            break  # found it
else:  # didn't find an ax we can reuse, so make a new one
    ax = fig.add_subplot(...)
```

File: lib/matplotlib/pyplot.py [hunk 7]
```diff
@@ -1228,11 +1249,29 @@ def subplot(*args, **kwargs):
...
+    # should re-use it
+    if ax is not None:
+        # sort out what projection and class would be created so we
+        # can tell if we should re-use an existing one
+        p_class, pkw = fig._process_projection_requirements(*args, **kwargs)
+        # if no projection or kwargs were passed, return the axes we
+        # found.
+        if projection is proj_placeholder and pkw == {}:
```
Comment (anntzer) [inline lib/matplotlib/pyplot.py:1260]: shouldn't this case be handled by the more general case below with `_projection_init == (Axes, {})`?

File: lib/matplotlib/pyplot.py [hunk 8]
```diff
@@ -1228,11 +1249,29 @@ def subplot(*args, **kwargs):
...
+        # otherwise we need to make a new one!
+        else:
+            ax = None
 
-    # If no existing axes match, then create a new one.
+    # if we do not want to re-use the existing Axes,
     if ax is None:
         ax = fig.add_subplot(*args, **kwargs)
 
+    fig.sca(ax)
```
Comment (anntzer) [inline lib/matplotlib/pyplot.py:1273]: was the call to sca just missing previously in the reuse axes case?

File: lib/matplotlib/pyplot.py (hunk 4 above)
Comment (anntzer) [reply lib/matplotlib/pyplot.py:1215]: That's a new dict that gets instantiated in the function call.

File: lib/matplotlib/pyplot.py (hunk 5 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:1208]: Because the user can palusibly pass in `None`, the goal here is to detect if the user passed us _anyting_ .  @anntzer 's suggest of calling this `unset` is better.

File: doc/api/next_api_changes/behavior/19438-TAC.rst (hunk 3 above)
Comment (tacaswell) [reply doc/api/next_api_changes/behavior/19438-TAC.rst:19]: no, this is a bug that goes back to 2.2.

File: lib/matplotlib/pyplot.py (hunk 6 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:1252]: Yup, I like that better.  I opted to do `ax=ax` rather than `pass` in the bodies of the if statements.

File: lib/matplotlib/pyplot.py (hunk 7 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:1260]: No, because we did not used to include the projection class in the key, only the kwargs, for the matching, and then did isintance checking to see if the Axes was a sub-class of the Axes we found.  The first condition is for the "just get me that Axes" logic, the second is for if the user passed anything that we need to resolve.

File: lib/matplotlib/pyplot.py (hunk 8 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:1273]: yes

File: doc/api/next_api_changes/behavior/19438-TAC.rst (hunk 3 above)
Comment (tacaswell) [reply doc/api/next_api_changes/behavior/19438-TAC.rst:19]: * atleast 2.2

File: lib/matplotlib/tests/test_pyplot.py [hunk 9]
```diff
@@ -161,3 +161,151 @@ def test_close():
...
+    assert ax3 is not ax2
+    assert ax2 not in plt.gcf().axes
+
+
+def test_subplot_polar_normalization():
+    ax1 = plt.subplot(111, projection='polar')
+    ax2 = plt.subplot(111, polar=True)
+    ax3 = plt.subplot(111, polar=True, projection='polar')
+    assert ax1 is ax2
+    assert ax1 is ax3
```
Comment (jklymak) [inline lib/matplotlib/tests/test_pyplot.py:295]: My opinion is still that a lot of the logic and copious comments in this PR is to satisfy this case.  While I freely admit this worked before, I still do not feel preserving this is worth all this extra logic.  The point here is to give the user a handle back to their original object.  Already `ax4=plt.subplot(111)` will do that for them, as well as calling it with the original kwargs.  I'd be shocked if we broke anyone by making the asserts fail, or if we did that they would not have a trivial fix to make it work again.

File: doc/api/next_api_changes/behavior/19438-TAC.rst (hunk 3 above)
Comment (tacaswell) [reply doc/api/next_api_changes/behavior/19438-TAC.rst:19]: but this text is missing a crucial half-sentence ....

Comment (lpsinger) [reply doc/api/next_api_changes/behavior/19438-TAC.rst]: Got it. Thanks!

File: lib/matplotlib/tests/test_pyplot.py [hunk 10]
```diff
@@ -161,3 +161,151 @@ def test_close():
...
         assert str(e) == "close() argument must be a Figure, an int, " \
                          "a string, or None, not <class 'float'>"
+
+
+def test_subplot_reuse():
+    ax1 = plt.subplot(121)
+    assert ax1 is plt.gca()
+    ax2 = plt.subplot(122)
+    assert ax2 is plt.gca()
+    ax1 = plt.subplot(121)
```
Comment (anntzer) [inline lib/matplotlib/tests/test_pyplot.py:171]: did you actually want to assert that gca() was the *original* ax1?

Comment (tacaswell) [reply lib/matplotlib/tests/test_pyplot.py]: fixed.

**Comment by tacaswell** [comment]
Simplified the implementation and line-wrapped the comments to 79 characters.

File: lib/matplotlib/pyplot.py (hunk 8 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:1260]: More verbosely, this behavior used to be implemented in the Figure methods but now needs to be done here instead.

**Review by anntzer — APPROVED** [review]

File: lib/matplotlib/pyplot.py [hunk 11]
```diff
@@ -1224,15 +1241,24 @@ def subplot(*args, **kwargs):
...
-         if hasattr(ax, 'get_subplotspec') and ax.get_subplotspec() == key),
-        None)
 
-    # If no existing axes match, then create a new one.
-    if ax is None:
+    for ax in fig.axes:
+        # if we found an axes at the position sort out if we can re-use it
+        if hasattr(ax, 'get_subplotspec') and ax.get_subplotspec() == key:
+            # if the user passed no kwargs, re-use
+            if kwargs == {}:
```
Comment (lpsinger) [inline lib/matplotlib/pyplot.py:1249]:
```suggestion
            if not kwargs:
```

but I know it's just a matter of style.

**Review by lpsinger — CHANGES_REQUESTED** [review]
Good news, astropy tests are still passing. Just spotted one typo, and suggested one very minor change that is a matter of style choice.

File: lib/matplotlib/pyplot.py [hunk 12]
```diff
@@ -1156,18 +1156,19 @@ def subplot(*args, **kwargs):
...
+    Axes being created.
 
-    In rare circumstances, `.Figure.add_subplot` may be called with a single
-    argument, a subplot axes instance already created in the
-    present figure but not in the figure's list of axes.
+    If *kwargs* are passed and there exists an Axes in the location
+    specified by *args*, the projection type is the same, and the
+    *kwargs* match with the existing Axes, then the existing Axes is
+    returned.  Otherwise a new Axes is created with the specified
+    parameters.  We save a reference to the *kwargs* which we us
```
[the remark this answers is not in the document]
Comment (lpsinger) [reply lib/matplotlib/pyplot.py:1167]: OK, I see this just got merged. I'll make a PR to fix the typo.

[the remark this answers is not in the document]
Comment (lpsinger) [reply lib/matplotlib/pyplot.py:1167]: See #19541.

File: lib/matplotlib/pyplot.py (hunk 11 above)
Comment (tacaswell) [reply lib/matplotlib/pyplot.py:1249]: I can see arguments either way here.  I want to say "If kwargs is equal to the empty dictionary" which is equivalent to "kwargs is falsy", but I think the `==` version communicates to the reader better the condition they should be thinking when the read it.
